scripts/upload-daemon/Main.hs

```diff
@@ -0,0 +1,128 @@
+import           Data.Maybe                                     (fromJust,
+                                                                 fromMaybe,
+                                                                 isJust,
+                                                                 listToMaybe)
+import           Network.Simple.TCP                             (HostPreference (Host),
+                                                                 SockAddr,
+                                                                 Socket,
+                                                                 closeSock,
+                                                                 recv, send,
+                                                                 serve)
+import           System.Environment                             (getArgs,
+                                                                 getEnv)
+import           System.Exit                                    (ExitCode (ExitSuccess))
+import           System.Metrics.Prometheus.Concurrent.RegistryT (registerCounter,
+                                                                 registerGauge,
+                                                                 runRegistryT)
+import           System.Metrics.Prometheus.Http.Scrape          (serveHttpTextMetricsT)
+import           System.Metrics.Prometheus.Metric.Counter       as C (Counter,
+                                                                      inc)
+import           System.Metrics.Prometheus.Metric.Gauge         as G (Gauge,
+                                                                      dec, inc)
+import           System.Metrics.Prometheus.MetricId             (addLabel,
+                                                                 fromList)
+import           System.Process                                 (readProcessWithExitCode)
+
+data StatisticsHandlers
+  = StatisticsHandlers
+  { requests :: Counter
+  , success  :: Gauge
+  , failure  :: Gauge
+  , running  :: Gauge
+  }
+
+type UploadTarget = String
+
+-- | First element of a list that is not Nothing
+firstJust :: [Maybe a] -> Maybe a
+firstJust = join . listToMaybe . Prelude.filter isJust
+
+
+-- | Upload target is taken either from first argument or from UPLOAD_TARGET env variable
+getUploadTarget :: IO (Maybe UploadTarget)
+getUploadTarget = firstJust <$> sequence
+  [ listToMaybe <$> getArgs
+  , optional $ getEnv "UPLOAD_TARGET"
+  ]
+
+-- | Get an environment variable with default fallback
+getEnvDefault :: String -> String -> IO String
+getEnvDefault var def = fromMaybe def <$> (optional $ getEnv var)
+
+-- | Receive data from `sock` until EOF or connection closure
+recvAll :: Socket -> IO (Maybe ByteString)
+recvAll sock = do
+  dat <- recv sock 1024
+  case dat of
+    Just str ->
+      if BS.length str == 1024
+      then (\rest -> dat <> rest) <$> recvAll sock
+      else return dat
+    Nothing -> return Nothing
```

Comment:
I guess it might be easier to rewrite this whole function using `unfoldM` from `monad-loops`. It will be as simple as `recvAll sock = mappend <$> unfoldM (recv sock 1024)`. I’m not really sure whether you actually want to distinguish between an empty bytestring and `Nothing`, this whole “this function returns `Nothing` when the connection is closed” thing doesn’t really make a lot of sense to me.

Reply:
Or you can just stick the bytestring length in front of it, which is, more or less, the only way to do this stuff reliably over TCP.
(Which reminds me of my suggestion to rewrite it in zmq, as it will handle the framing for you and is, in general, easier to use and more reliable.)

Reply:
> It will be as simple as recvAll sock = mappend <$> unfoldM (recv sock 1024)

No, this is not the same implementation. We need to check if the length of the ByteString is equal to the limit of characters. Your implemetation will block until you kill `nc` (thus closing the socket).

Reply:
> recvAll sock = mappend <$> unfoldM (recv sock 1024)

This is not the same implementation, it doesn't work (because it doesn't check if `recv` returned because of EOF or byte limit, and we only need to continue receving in the latter case).

> Or you can just stick the bytestring length in front of it, which is, more or less, the only way to do this stuff reliably over TCP.

Eww. With nc?